idx/metric_definition.go

```diff
@@ -0,0 +1,351 @@
+func (md *MetricDefinition) NameWithTags() string {
+	bld := strings.Builder{}
+
+	md.Name.string(&bld)
+	sort.Sort(TagKeyValues(md.Tags))
+	for _, tag := range md.Tags {
+		if tag.Key == "name" {
+			continue
+		}
+		bld.WriteString(";")
+		bld.WriteString(tag.String())
+	}
+	return bld.String()
+}
+
+// SetMType translates a string into a uint8 which is used to store
+// the actual metric type. Valid values are 'gauge', 'rate', 'count',
+// 'counter', and 'timestamp'.
+func (md *MetricDefinition) SetMType(mtype string) {
+	switch mtype {
+	case "gauge":
+		md.mtype = MTypeGauge
+	case "rate":
+		md.mtype = MTypeRate
+	case "count":
+		md.mtype = MTypeCount
+	case "counter":
+		md.mtype = MTypeCounter
+	case "timestamp":
+		md.mtype = MTypeTimestamp
+	default:
+		// for values "" and other unknown/corrupted values
+		md.mtype = MTypeUndefined
+	}
+}
+
+// Mtype returns a string version of the current MType
+func (md *MetricDefinition) Mtype() string {
+	switch md.mtype {
+	case MTypeGauge:
+		return "gauge"
+	case MTypeRate:
+		return "rate"
+	case MTypeCount:
+		return "count"
+	case MTypeCounter:
+		return "counter"
+	case MTypeTimestamp:
+		return "timestamp"
+	default:
+		// case of MTypeUndefined and also default for unknown/corrupted values
+		return ""
+	}
+}
+
+// SetUnit takes a string, interns it in an object store
+// and then uses it to store the unit.
+func (md *MetricDefinition) SetUnit(unit string) {
+	sz, err := IdxIntern.AddOrGetSzNoCprsn([]byte(unit))
+	if err != nil {
+		log.Errorf("idx: Failed to intern Unit %v. %v", unit, err)
+		md.Unit = unit
+	}
+	md.Unit = sz
+}
+
+// SetMetricName interns the MetricName in an
+// object store and stores the addresses of those strings
+// in MetricName.nodes
+func (md *MetricDefinition) SetMetricName(name string) {
+	nodes := strings.Split(name, ".")
+	md.Name.nodes = make([]uintptr, len(nodes))
+	for i, node := range nodes {
+		// TODO: add error checking? Fail somehow
+		nodePtr, err := IdxIntern.AddOrGet([]byte(node))
+		if err != nil {
+			log.Errorf("idx: Failed to intern word in MetricName: %v, %v", node, err)
+		}
+		md.Name.nodes[i] = nodePtr
+	}
+}
+
+// SetTags takes a []string which should contain Key/Value pairs
+// in the form of 'key=value'. It splits up the Key and Value for each
+// item, interns them in the object store, and creates a TagKeyValue
+// for them. It then stores all of these in Tags.
+//
+// The items in the input argument should not contain ';'. Each item
+// is a separate Key/Value pair. Do not combine multiple Key/Value pairs
+// into a single index in the []string.
+func (md *MetricDefinition) SetTags(tags []string) {
+	md.Tags = make([]TagKeyValue, len(tags))
+	sort.Strings(tags)
+	for i, tag := range tags {
+		if strings.Contains(tag, ";") {
+			invalidTag.Inc()
+			log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+			continue
+		}
+		splits := strings.Split(tag, "=")
+		if len(splits) < 2 {
```

Comment:
```suggestion
		if len(splits) != 2 {
```

Reply:
actually, better would be to not allocate a new slice. just count the number of `=` and get the position of the character, you can then index into the input slice